### PR TITLE
upt-invariant-mass_l1t_11_2_0

### PR DESCRIPTION
**Pull request by cavana**
#### PR description:

This PR provides a software implementation of the invariant mass correlation condition involving unconstrained pT for displaced muons in the uGT emulator. 

No changes in output are expected.

#### PR validation:

This PR has been validated by the following:
(1) all code changes have been confirmed to compile correctly using l1t-integration-CMSSW_11_2_0 with tag l1t-integration-v105.3
(2) all code changes have been confirmed to execute properly by producing test vectors
(3) confirming that the software implementation is correct by comparing the emulator results in the test vectors with the uGT firmware.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

This particular PR is not a backport

Before submitting your pull requests, make sure you followed this checklist:
- verify that the PR is really intended for the chosen branch
- verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)

### CODE REVIEW

File: L1Trigger/L1TGlobal/src/CorrCondition.cc [hunk 1]
```diff
@@ -321,6 +321,14 @@ const bool l1t::CorrCondition::evaluateCondition(const int bxEval) const {
   int etBin1 = 0;
   double et1Phy = 0.;
 
+  // Added by R.Cavanaugh for displaced muons
```
Comment (rekovic) [inline L1Trigger/L1TGlobal/src/CorrCondition.cc]: Please remove meaningless comments for understanding the code.  Git versioning system keeps track of authors of commits.

File: L1Trigger/L1TGlobal/src/CorrCondition.cc [hunk 2]
```diff
@@ -370,13 +378,15 @@ const bool l1t::CorrCondition::evaluateCondition(const int bxEval) const {
         phiIndex0 = (candMuVec->at(cond0bx, obj0Index))->hwPhiAtVtx();  //(*candMuVec)[obj0Index]->phiIndex();
         etaIndex0 = (candMuVec->at(cond0bx, obj0Index))->hwEtaAtVtx();
         etIndex0 = (candMuVec->at(cond0bx, obj0Index))->hwPt();
+	uptIndex0 = (candMuVec->at(cond0bx, obj0Index))->hwPtUnconstrained(); // Added by R.Cavanaugh for displaced muons
```
Comment (rekovic) [inline L1Trigger/L1TGlobal/src/CorrCondition.cc]: Please remove meaningless comments.  Git versioning system keeps track of authors of commits.

File: L1Trigger/L1TGlobal/src/CorrCondition.cc [hunk 3]
```diff
@@ -646,13 +659,15 @@ const bool l1t::CorrCondition::evaluateCondition(const int bxEval) const {
           phiIndex1 = (candMuVec->at(cond1bx, obj1Index))->hwPhiAtVtx();  //(*candMuVec)[obj0Index]->phiIndex();
           etaIndex1 = (candMuVec->at(cond1bx, obj1Index))->hwEtaAtVtx();
           etIndex1 = (candMuVec->at(cond1bx, obj1Index))->hwPt();
+	  uptIndex1 = (candMuVec->at(cond1bx, obj1Index))->hwPtUnconstrained(); // Added by R.Cavanaugh for displaced muons
```
Comment (rekovic) [inline L1Trigger/L1TGlobal/src/CorrCondition.cc]: Please remove meaningless comments  for understanding the code.  Git versioning system keeps track of authors of commits.

File: L1Trigger/L1TGlobal/src/CorrCondition.cc [hunk 4]
```diff
@@ -646,13 +659,15 @@ const bool l1t::CorrCondition::evaluateCondition(const int bxEval) const {
...
           chrg1 = (candMuVec->at(cond1bx, obj1Index))->hwCharge();
           etaBin1 = etaIndex1;
           if (etaBin1 < 0)
             etaBin1 = m_gtScales->getMUScales().etaBins.size() + etaBin1;
           //		   LogDebug("L1TGlobal") << "Muon phi" << phiIndex1 << " eta " << etaIndex1 << " etaBin1 = " << etaBin1  << " et " << etIndex1 << std::endl;
 
           etBin1 = etIndex1;
+	  uptBin1 = uptIndex1;   // Added by R.Cavanaugh for displaced muons
```
Comment (rekovic) [inline L1Trigger/L1TGlobal/src/CorrCondition.cc]: Please remove meaningless comments.  Git versioning system keeps track of authors of commits.

File: L1Trigger/L1TGlobal/src/CorrCondition.cc [hunk 5]
```diff
@@ -1149,7 +1166,7 @@ const bool l1t::CorrCondition::evaluateCondition(const int bxEval) const {
         }
       }
 
-      if (corrPar.corrCutType & 0x8 || corrPar.corrCutType & 0x10) {
+      if (corrPar.corrCutType & 0x8 || corrPar.corrCutType & 0x10 || corrPar.corrCutType & 0x40) { // R. Cavanaugh added 0x40 for massUpt
```
Comment (rekovic) [inline L1Trigger/L1TGlobal/src/CorrCondition.cc]: Please remove meaningless comments. Git versioning system keeps track of authors of commits.

File: L1Trigger/L1TGlobal/src/CorrCondition.cc [hunk 6]
```diff
@@ -1159,6 +1176,8 @@ const bool l1t::CorrCondition::evaluateCondition(const int bxEval) const {
         if (corrPar.corrCutType & 0x10)
           coshDeltaEtaPhy = 1.;
         double massSqPhy = et0Phy * et1Phy * (coshDeltaEtaPhy - cosDeltaPhiPhy);
+	if( corrPar.corrCutType & 0x40 ) // Added by R.Cavanaugh for displaced muons
```
Comment (rekovic) [inline L1Trigger/L1TGlobal/src/CorrCondition.cc]: Please remove meaningless comments. Git versioning system keeps track of authors of commits.

File: L1Trigger/L1TGlobal/src/CorrCondition.cc [hunk 7]
```diff
@@ -666,7 +681,9 @@ const bool l1t::CorrCondition::evaluateCondition(const int bxEval) const {
           eta1Phy = 0.5 * (binEdges.second + binEdges.first);
           binEdges = m_gtScales->getMUScales().etBins.at(etBin1);
           et1Phy = 0.5 * (binEdges.second + binEdges.first);
-
+	  // Added by R.Cavanaugh for displaced muons
```
Comment (rekovic) [inline L1Trigger/L1TGlobal/src/CorrCondition.cc]: Please remove meaningless comments. Git versioning system keeps track of authors of commits.

File: L1Trigger/L1TGlobal/src/CorrCondition.cc [hunk 8]
```diff
@@ -390,6 +400,9 @@ const bool l1t::CorrCondition::evaluateCondition(const int bxEval) const {
         eta0Phy = 0.5 * (binEdges.second + binEdges.first);
         binEdges = m_gtScales->getMUScales().etBins.at(etBin0);
         et0Phy = 0.5 * (binEdges.second + binEdges.first);
+	// Added by R.Cavanaugh for displaced muons
```
Comment (rekovic) [inline L1Trigger/L1TGlobal/src/CorrCondition.cc]: Please remove meaningless comments. Git versioning system keeps track of authors of commits.

File: L1Trigger/L1TGlobal/src/CorrCondition.cc [hunk 9]
```diff
@@ -370,13 +378,15 @@ const bool l1t::CorrCondition::evaluateCondition(const int bxEval) const {
...
         etIndex0 = (candMuVec->at(cond0bx, obj0Index))->hwPt();
+	uptIndex0 = (candMuVec->at(cond0bx, obj0Index))->hwPtUnconstrained(); // Added by R.Cavanaugh for displaced muons
         chrg0 = (candMuVec->at(cond0bx, obj0Index))->hwCharge();
         int etaBin0 = etaIndex0;
         if (etaBin0 < 0)
           etaBin0 = m_gtScales->getMUScales().etaBins.size() + etaBin0;  //twos complement
         //		LogDebug("L1TGlobal") << "Muon phi" << phiIndex0 << " eta " << etaIndex0 << " etaBin0 = " << etaBin0  << " et " << etIndex0 << std::endl;
 
         etBin0 = etIndex0;
+	uptBin0 = uptIndex0;   // Added by R.Cavanaugh for displaced muons
```
Comment (rekovic) [inline L1Trigger/L1TGlobal/src/CorrCondition.cc]: Please remove meaningless comments. Git versioning system keeps track of authors of commits.

File: L1Trigger/L1TGlobal/src/CorrCondition.cc [hunk 10]
```diff
@@ -370,13 +378,15 @@ const bool l1t::CorrCondition::evaluateCondition(const int bxEval) const {
         phiIndex0 = (candMuVec->at(cond0bx, obj0Index))->hwPhiAtVtx();  //(*candMuVec)[obj0Index]->phiIndex();
         etaIndex0 = (candMuVec->at(cond0bx, obj0Index))->hwEtaAtVtx();
         etIndex0 = (candMuVec->at(cond0bx, obj0Index))->hwPt();
+	uptIndex0 = (candMuVec->at(cond0bx, obj0Index))->hwPtUnconstrained(); // Added by R.Cavanaugh for displaced muons
         chrg0 = (candMuVec->at(cond0bx, obj0Index))->hwCharge();
         int etaBin0 = etaIndex0;
         if (etaBin0 < 0)
           etaBin0 = m_gtScales->getMUScales().etaBins.size() + etaBin0;  //twos complement
         //		LogDebug("L1TGlobal") << "Muon phi" << phiIndex0 << " eta " << etaIndex0 << " etaBin0 = " << etaBin0  << " et " << etIndex0 << std::endl;
```
Comment (rekovic) [inline L1Trigger/L1TGlobal/src/CorrCondition.cc]: Please remove commented out code.  Or uncomment it if useful.

File: L1Trigger/L1TGlobal/plugins/TriggerMenuParser.cc [hunk 11]
```diff
@@ -2582,9 +2597,17 @@ bool l1t::TriggerMenuParser::parseCorrelation(tmeventsetup::esCondition corrCond
           cutType = cutType | 0x8;
           //std::cout << "CCLA running Invarient mass cutType= " << cutType << std::endl;
```
Comment (rekovic) [inline L1Trigger/L1TGlobal/plugins/TriggerMenuParser.cc]: Please remove commented code.

**Review by rekovic — CHANGES_REQUESTED** [review]

File: L1Trigger/L1TGlobal/src/CorrCondition.cc (hunk 1 above)
Comment (cavana) [reply L1Trigger/L1TGlobal/src/CorrCondition.cc]: @rekovic 
Many thanks for the helpful request to remove comments.  Done.

File: L1Trigger/L1TGlobal/src/CorrCondition.cc (hunk 2 above)
Comment (cavana) [reply L1Trigger/L1TGlobal/src/CorrCondition.cc]: @rekovic 
Many thanks for the helpful request to remove comments.  Done.

File: L1Trigger/L1TGlobal/src/CorrCondition.cc (hunk 3 above)
Comment (cavana) [reply L1Trigger/L1TGlobal/src/CorrCondition.cc]: @rekovic 
Many thanks for the helpful request to remove comments.  Done.

File: L1Trigger/L1TGlobal/src/CorrCondition.cc (hunk 4 above)
Comment (cavana) [reply L1Trigger/L1TGlobal/src/CorrCondition.cc]: @rekovic 
Many thanks for the helpful request to remove comments.  Done.

File: L1Trigger/L1TGlobal/src/CorrCondition.cc (hunk 5 above)
Comment (cavana) [reply L1Trigger/L1TGlobal/src/CorrCondition.cc]: @rekovic 
Many thanks for the helpful request to remove comments.  Done.

File: L1Trigger/L1TGlobal/src/CorrCondition.cc (hunk 7 above)
Comment (cavana) [reply L1Trigger/L1TGlobal/src/CorrCondition.cc]: @rekovic 
Many thanks for the helpful request to remove comments.  Done.

File: L1Trigger/L1TGlobal/src/CorrCondition.cc (hunk 8 above)
Comment (cavana) [reply L1Trigger/L1TGlobal/src/CorrCondition.cc]: @rekovic 
Many thanks for the helpful request to remove comments.  Done.

File: L1Trigger/L1TGlobal/src/CorrCondition.cc (hunk 9 above)
Comment (cavana) [reply L1Trigger/L1TGlobal/src/CorrCondition.cc]: @rekovic 
Many thanks for the helpful request to remove comments.  Done.

File: L1Trigger/L1TGlobal/src/CorrCondition.cc (hunk 10 above)
Comment (cavana) [reply L1Trigger/L1TGlobal/src/CorrCondition.cc]: @rekovic 
Many thanks for the helpful request to remove comments.  Done.

File: L1Trigger/L1TGlobal/plugins/TriggerMenuParser.cc (hunk 11 above)
Comment (cavana) [reply L1Trigger/L1TGlobal/plugins/TriggerMenuParser.cc]: @rekovic 
Many thanks for the helpful request to remove comments.  Done.

**Comment by cavana** [comment]
Please do not accept/merge this PR until additional checks are performed to (re-)verify agreement with firmware

**Comment by rekovic** [comment]
@cavana this is merged now. If you have additional fixes, please open a new PR.

**Pull request merged**